pkg/agent/plugin/workloadattestor/systemd/systemd_posix.go

```diff
@@ -0,0 +1,116 @@
+//go:build !windows
+// +build !windows
+
+package systemd
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/godbus/dbus/v5"
+	"github.com/hashicorp/go-hclog"
+	workloadattestorv1 "github.com/spiffe/spire-plugin-sdk/proto/spire/plugin/agent/workloadattestor/v1"
+	"github.com/spiffe/spire/pkg/common/catalog"
+)
+
+func builtin(p *Plugin) catalog.BuiltIn {
+	return catalog.MakeBuiltIn(pluginName,
+		workloadattestorv1.WorkloadAttestorPluginServer(p),
+	)
+}
+
+type unitInfo interface {
+	Id() (string, error)
+	FragmentPath() (string, error)
+}
+
+type DBusUnitObject struct {
+	dbus.BusObject
+}
+
+func getStringProperty(obj DBusUnitObject, prop string) (string, error) {
+	propVariant, err := obj.GetProperty("org.freedesktop.systemd1.Unit." + prop)
+	if err != nil {
+		return "", err
+	}
+	propVal, ok := propVariant.Value().(string)
+	if !ok {
+		return "", fmt.Errorf("Returned value for %v was not a string: %v", prop, propVariant.String())
+	}
+	return propVal, nil
+}
+
+func (obj DBusUnitObject) Id() (string, error) {
+	return getStringProperty(obj, "Id")
+}
+
+func (obj DBusUnitObject) FragmentPath() (string, error) {
+	return getStringProperty(obj, "FragmentPath")
+}
+
+type Plugin struct {
+	workloadattestorv1.UnsafeWorkloadAttestorServer
+
+	log hclog.Logger
+
+	// hook for tests
+	getUnitInfo func(ctx context.Context, pid uint) (unitInfo, error)
+}
+
+func getSystemdUnitInfo(ctx context.Context, pid uint) (unitInfo, error) {
+	conn, err := dbus.SystemBus()
```

Comment:
This connection should be closed (using a defer) to ensure timely releasing of resources.